### PR TITLE
Make null param error more informative (#809)

### PR DESCRIPTION
**Pull request by ulfryk**

### CODE REVIEW

File: modules/core/src/main/scala/doobie/util/put.scala [hunk 1]
```diff
@@ -56,25 +56,37 @@ sealed abstract class Put[A](
...
   def unsafeSetNullable(ps: PreparedStatement, n: Int, oa: Option[A]): Unit =
     oa match {
-      case Some(a) => unsafeSetNonNullable(ps, n, a)
+      case Some(a) => unsafeSetWhatShouldNotBeNull(ps, n, a, expectedOptionalParamAt(n))
       case None    => unsafeSetNull(ps, n)
     }
 
+  private def unsafeSetWhatShouldNotBeNull(ps: PreparedStatement, n: Int, a: A, message: String): Unit =
+    if (a == null) sys.error(message)
+    else put.fi.apply(ps, n, (put.k(a)))
```
Comment (jatcwang) [inline modules/core/src/main/scala/doobie/util/put.scala:69]: In this case we're creating a new string for every value we set, which is probably not ideal for performance.
We probably want to only construct the message in the `if (a == null)` branch. To simplify things, I think it's probably not necessary to differentiate between `null` and `Option(null)` (though I appreciate the thoughtfulness behind it :slightly_smiling_face: )

Comment (ulfryk) [reply modules/core/src/main/scala/doobie/util/put.scala:69]: @jatcwang I have updated the code. It was the `n` param info that caused creating new string for every value we set. Now both strings are persisted and we keep the `null` vs `Option(null)` distinction :)

Comment (ulfryk) [reply modules/core/src/main/scala/doobie/util/put.scala:69]: Or do you think super simple would be even better? Like this --> https://github.com/typelevel/doobie/pull/2142

Comment (ulfryk) [reply modules/core/src/main/scala/doobie/util/put.scala:69]: Or with message constructed only in the `if (a == null)` branch and full information --> #2143

**Pull request closed without merging**